app/models/concerns/warehouse_reports/publish.rb

```diff
@@ -0,0 +1,158 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module WarehouseReports::Publish
+  extend ActiveSupport::Concern
+  include WarehouseReports::S3Toolset
+  included do
+    def self.published(path)
+      GrdaWarehouse::PublishedReport.where(report_type: sti_name, path: path, state: 'published').first
+    end
+
+    # Override as necessary
+    def publish_files
+      [
+        {
+          name: 'index.html',
+          content: -> { as_html },
+          type: 'text/html',
+        },
+      ]
+    end
+
+    def publish_warning
+      previously_published = self.class.published(path)
+      return nil if previously_published.blank?
+
+      "Publishing this version of the #{instance_title} will remove any similar previously published version regardless of who published it.  The currently published version is from #{previously_published.updated_at.to_date}.  Are you sure you want to un-publish the previous version and publish this version?"
+    end
+
+    def un_publish_warning
+      'Un-publishing this report will prevent anyone from accessing the published version of this report.  It will delete any existing web pages that have been published, and if the pages are embedded in a public website, those pages may not function correctly.'
+    end
+
+    # Your report is required to define public_s3_directory
+    # private def public_s3_directory
+    #   # This should return a string that represends the unique path for aLl reports of this type
+    #   'path-to-report'
+    # end
```

Comment:
Maybe a `raise` here in case it gets forgotten?

Reply:
So, maybe, the issue is that often this concern is included _after_ the definition of the method and I didn't want to override it.